dicom_utils/volume.py

```diff
@@ -2,28 +2,87 @@
 # -*- coding: utf-8 -*-
 
 from abc import ABC, abstractmethod
-from typing import Sized, SupportsIndex, TypeVar
+from copy import deepcopy
+from itertools import islice
+from typing import Iterator, Optional, Sized, SupportsIndex, SupportsInt, Tuple, TypeVar, Union, overload
+
+import numpy as np
+from pydicom import Dataset
+from pydicom.encaps import encapsulate, generate_pixel_data_frame
 
 
 T = TypeVar("T", bound=SupportsIndex)
+U = TypeVar("U", bound=Dataset)
 
 
 class VolumeHandler(ABC):
     r"""Base class for classes that manipulate 3D Volumes"""
 
     @abstractmethod
+    def get_indices(self, total_frames: Optional[int]) -> Tuple[int, int, int]:
+        ...
+
+    @overload
     def __call__(self, x: T) -> T:
         ...
 
+    @overload
+    def __call__(self, x: U) -> U:
+        ...
+
+    def __call__(self, x: Union[T, U]) -> Union[T, U]:
+        if isinstance(x, Dataset):
+            return self.slice_dicom(x)
+        else:
+            return self.slice_array(x)
+
+    def slice_array(self, x: T) -> T:
+        r"""Slices an array input according to :func:`get_indices`"""
+        num_frames = len(x) if isinstance(x, Sized) else None
+        start, stop, stride = self.get_indices(num_frames)
+        result = x[start:stop:stride]  # type: ignore
+        return result
+
+    def slice_dicom(self, dcm: U) -> U:
+        r"""Slices a DICOM object input according to :func:`get_indices`.
+
+        .. note:
+            Unlike :func:`slice_array`, this function can perform slicing on compressed DICOMs
+            with out needing to decompress all frames. This can provide a substantial performance gain.
+
+        """
+        # copy dicom and read key tags
+        dcm = deepcopy(dcm)
+        num_frames: Optional[SupportsInt] = dcm.get("NumberOfFrames", None)  # type: ignore
+        num_frames = int(num_frames) if num_frames is not None else None
+        is_compressed: bool = dcm.file_meta.TransferSyntaxUID.is_compressed  # type: ignore
+
+        start, stop, stride = self.get_indices(num_frames)
+
+        # read data
+        if is_compressed:
+            all_frames: Iterator = generate_pixel_data_frame(dcm.PixelData, num_frames)  # type: ignore
```

Comment:
I suggest renaming `all_frames` to `frame_iterator` or something similar and then the `# type: ignore` can be dropped.